### ISSUE TITLE
add `namespaces` functionality

### ISSUE THREAD

**Issue by ivanhofer** [ivanhofer/typesafe-i18n#70]
Namespaces would be a native option to load only selected parts of the translations. Useful for bigger applications so not all messages need to load the same time e.g. translation that are only needed in the settings panel are not needed on the login page.

You would define your translations something like:

```typescript
// index.ts
import type { Translation } from '../i18n-types'
import { workspace } from 'typesafe-i18n'

const en: Translation = {
   login: workspace('login')
   settings: workspace('settings')
   dashboard: {
      // ...
   }
}

export default en

// login.ts
import type { LoginTranslation } from '../i18n-types'

const login: LoginTranslation = {
   label_password: 'Password:',
   label_username: 'Username:',
   // ...
}

export default login

// settings.ts
import type { SettingsTranslation } from '../i18n-types'

const settings: SettingsTranslation = {
   email: 'Email preferences:',
   // ...
}

export default settings
```

where `workspace` would be a function that loads the workspace via an async import

```typescript

const loadedWorkspaces = { }

export const workspace = async (workspaceToLoad: string) => {
  const ws = loadedWorkspaces[workspaceToLoad]
   if (ws) return ws

   loadWorkspace(workspaceToLoad)

   return { } // a dummy-workspace
}

export const loadWorkspace = async (workspaceToLoad: string) => {
   loadedWorkspaces[workspaceToLoad] = (await import(`./${workspaceToLoad}`)).default
}
```

Because workspaces would be loaded asynchronously, we would make sure that calling `LL.login.label_password()` would return the translated value: How can we do that?
 - call somewhere a `loadWorkspace('login')` function
     the function would need to get called before the actual translation call
 - show an empty string until the data finishes loading in background
     since the whole translation object is reactive, we can initialize the loading and as soon as the translations are back, we can update the object that holds all workspaces

A developer then can decide if he manually loads the correct workspace somewhere in the code (e.g. triggered by a certain route) or the user should see blanks until translations are loaded. Also a method to check if a workspace is currently loading will be exposed so a loading spinner can be shown.

It should be possible to nest workspaces as often an user wants.

What needs to be done:
 - [ ] create custom types that only include part of the workspace translations
 - [ ] make sure all translations define the part of the subtree as a workspace
 - [ ] implement reactivity for loading workspaces asynchronously
 - [ ] function to check if workspace is currently loading

**Comment by mullwaden** [comment]
Kind of related feature that might be interesting. In a very large project, having all of the copy in one place becomes challenging. Right now (I could be wrong) the setup forces a structure like:

```
src/notifications/published.ts
src/i18n/translations/en/published_copy.ts
src/i18n/translations/de/published_copy.ts
```

whereas something like this, allowing the copy to be colocated with the component/function could be very helpful to avoid jumping around in the codebase:

```
src/notifications/published.ts
src/notifications/published_copy.ts  (perhaps with sub-exports like   export const en = { } )
```

**Comment by ivanhofer** [comment]
I'm not sure if I got it correct.

You can do that already, but you would need to import these objects inside the `src/i18n/[locale]/index.ts` file.
To simplify the generator, it uses the `index.ts` file's default export as the entry point to all your locales. What you then do inside that file is up to you. You can import anything that lays inside the root of your project.

You can have a file `src/notifications/published.ts` that exports part of a dictionary in multiple languages like
```typescript
export const notifications_en = {
  prop1: 'This is a test'
}

export const notifications_de = {
  prop1: 'Das ist ein test'
}
```

and then reference it inside `src/i18n/en/index.ts` like
```typescript
import { notifications_en } from 'src/notifications/published.ts'

const en: BaseTranslation = {
   ...notifications_en
}
export default en
```
and inside `src/i18n/de/index.ts` like
```typescript
import { notifications_de } from 'src/notifications/published.ts'

const de: Translation = {
   ...notifications_de
}
export default de
```

Is this what you want?

**Comment by ivanhofer** [comment]
@mullwaden did you see my comment above?

**Comment by mullwaden** [comment]
yes! I ended up putting everything in the translations folder though. It works better with the automatic updates of the types.

One idea would be to go the route that jest, storybook and less go, where they automatically pick up files, e.g.
myClass.ts
myClass.test.ts

Imagine then you could do something like

```
import translation from "./myClass.translation";

const title = translation.title();
```

It can be done, but it would require some interesting webpack shenanigans :)

**Comment by ivanhofer** [comment]
@mullwaden sounds interesting. Pleas open a [`discussion`](https://github.com/ivanhofer/typesafe-i18n/discussions) with some more details. Maybe it is a topic for the future.
